### ISSUE TITLE
Need confirmation for some unpatched CVE in Curl

### ISSUE THREAD

**Issue by the-Chain-Warden-thresh** [premake/premake-core#2156]
I'm cloning this repo to make some modifications to customize. However, I've noticed that some CVEs which were confirmed and fixed by [curl](https://github.com/curl/curl) do not get patched in this repo. To enhance the availability of my project as far as possible, I will appreciate it if any of the CVE below do exist in this repo as well, so that I can fix these security issue myself by applying the corresponding patch.
Here are the CVEs I found in this repo unpatched, but get fixed in [curl](https://github.com/curl/curl):

CVE-2018-1000121 in [contrib/curl/lib/openldap.c](https://github.com/premake/premake-core/blob/master/contrib/curl/lib/openldap.c#L414)'s function `static ssize_t ldap_recv(struct connectdata *conn, int sockindex, char *buf, size_t len, CURLcode *err)`, with [patch](https://github.com/curl/curl/commit/9889db043393092e9d4b5a42720bba0b3d58deba) here for your reference.

CVE-2018-16890 in [contrib/curl/lib/vauth/ntlm.c](https://github.com/premake/premake-core/blob/master/contrib/curl/lib/vauth/ntlm.c#L167)'s function `static CURLcode ntlm_decode_type2_target(struct Curl_easy *data,      unsigned char *buffer, size_t size,
                                         struct ntlmdata *ntlm)`, with [patch](https://github.com/curl/curl/commit/b780b30d1377adb10bbe774835f49e9b237fb9bb) here for your reference.

CVE-2018-1000122 in [contrib/curl/lib/transfer.c](https://github.com/premake/premake-core/blob/master/contrib/curl/lib/transfer.c#L389)'s function `static CURLcode readwrite_data(struct Curl_easy *data,
                               struct connectdata *conn,
                               struct SingleRequest *k,
                               int *didwhat, bool *done,
                               bool *comeback)`, with [patch](https://github.com/curl/curl/commit/d52dc4760f6d9ca1937eefa2093058a952465128) here for your reference.

**Comment by nickclark2016** [comment]
Instead of manually applying patches, we'd rather just update our build of curl.

**Comment by nekobbbbbbit** [comment]
I've complained about this before somewhere in #2057 ... I've discussed about ways to fetch an up-to-date curl build, and i also didn't mention pulling curl from source. 

What should be happening is that Curl gets pulled and either patches get applied (per OS?) or defer to OS libraries (on linux, bsd, or even windows now that Curl seems bundled in). Curl is designed to build on quite literally everything and the reasoning for maintaining custom patches seems strange to me.

I even had to manually patch some pieces on openbsd iirc, which obviously goes to show that this isn't really a portable solution anyway. It should at _least_ fall back to pkg-config or something. Many CMake projects even do that

I want to tag @bagder, I think i asked his input in IRC a long long time ago on bundling old Curl in this project. Maybe he will have better resourcing of what's going on here.

**Comment by bagder** [comment]
I don't have much to contribute when it comes how you should handle curl in your project. I'm convinced you know much more about that and your project's specifics than I do.

In curl, we document every past curl CVE with lots of details here: https://curl.se/docs/security.html

**Comment by nekobbbbbbit** [comment]
> I don't have much to contribute when it comes how you should handle curl in your project. I'm convinced you know much more about that and your project's specifics than I do.

@bagder https://premake.github.io/docs/http.download - Seems like a fairly common endpoint.

Generally speaking, it's absolutely possible for someone to hijack their server if they offer downloads to something, no? At least with such an old version of libcurl, do you know any sort of specific remote-execution since (2018) then, or any that are particularly easy to pull off? I probably, and maybe others, aren't worried about DoS attacks or assert-like crashes. The real issue can be remote attacks.

Most importantly, Look at these **very small** commits / patches (peep the local curl changes) to the actual Curl distribution, do they come off a bit strange or hacky?

- [Patch 1](https://github.com/premake/premake-core/commit/31e3f5d16939336b5a940a5014a37bb9e589e9f8)
- [Patch 2](https://github.com/premake/premake-core/commit/755a7e6d66f9f42f0e30938406f2853ae2aa4dfb#diff-8e012f804d713dbbaf93061ab5a5c31381f9ad243635788cdb04d2c9dd96dc70)
- [Patch 3](https://github.com/premake/premake-core/commit/d1bce0c95e16151101c4d62295aacf608c1e0cb4)
- [Patch 4](https://github.com/premake/premake-core/commit/fd8bfef352e15066d0711364bb53f73eb3c083b9#diff-8e012f804d713dbbaf93061ab5a5c31381f9ad243635788cdb04d2c9dd96dc70)
- [Patch 5 (this one is a bit larger than others)](https://github.com/premake/premake-core/commit/907c83d840bc911bc34dfb01a953df36bc730f8d)
- [Patch 6](https://github.com/premake/premake-core/commit/0884f50742a5621781f3f48e80cd86d111b2ae59)
- [Patch 7](https://github.com/premake/premake-core/commit/31e3f5d16939336b5a940a5014a37bb9e589e9f8)

> A better package management mechanism can definitely minimize similar issues like this as much as possible, but I'd rather manually apply patches so that I can start my own work as soon as possible.

@the-Chain-Warden-thresh Generally speaking, everything _local_ should be done in a Patch file (or several patch files), which gets applied to the latest stable release of Curl. Instead, we're just bloating up a Git repository with a copy of Curl (and other repositories). It just seems quite wrong.

**Comment by nickclark2016** [comment]
Personally, I'd prefer to either be on the latest version of libcurl (and automatically update somehow) or use the OS built-in libraries. That said, I do not have the time to investigate and implement either of these approaches. If there are security vulnerabilities and patching would be an immediate stopgap, I'd happily review patches to merge them in. I'd also happily other changes that more security minded people (like yourselves) could contribute from libcurl to the repo. I can honestly say that security is not in my core skillset, so I'll be happy to review any changes or recommendations made.

**Comment by nickclark2016** [comment]
@starkos @samsinsane Please take a look at this. I give it my approval, but I'd prefer to have one of you also on the thread for a greenlight.

**Comment by samsinsane** [comment]
There's a lot of things to respond to here.

### Updating dependencies

Yeah, sounds good. However, not all users will be happy with this happening. For example, when we moved from Lua 5.1 to 5.3, this broke compatibility with LuaJIT which caused issues for some users. So, not everyone will be interested in updating/security at all costs.

### Using system libraries

This is at odds with another issue (#632). Since binaries are made available on the GitHub releases, it makes more sense to continue to bundle the dependencies as we currently do. Providing an option for users to build with system libraries is fine, but it shouldn't come at the detriment of the existing workflow.

### Downloading the latest versions of dependencies

Personally, not interested in this. The current set of prerequisites for building Premake is Visual Studio or make+GCC (or drop-in replacements). Personally, this lower barrier to entry was one of the reasons I got into working on Premake in the first place. I didn't need to setup a bunch of random dependencies just to build Premake, or to extend it. This made it easier to just get on with what I wanted to do, which lead to me becoming a maintainer of the project.

### Using submodules instead of copying code

I'm in favour of this, with the caveat of the above - it shouldn't increase the prerequisites for building Premake.

---

Having said all of that, I think it's important to have @starkos provide his input before changing how the project is structured. I know he hasn't had time to respond to issues recently but I'm not a fan of changing the project while he's unable to approve or provide input.

**Comment by samsinsane** [comment]
No, there's been no progress on this. Any progress will be noted in this issue.
